cpp/src/arrow/acero/source_node.cc

```diff
@@ -263,6 +263,22 @@ struct SourceNode : ExecNode, public TracedNode {
     to_finish.MarkFinished();
   }
 
+  Status StopProducing() override {
+    // GH-35837: ensure node is not paused
+    Future<> to_finish;
+    {
+      std::lock_guard<std::mutex> lg(mutex_);
+      if (!backpressure_future_.is_finished()) {
```

Comment:
What does this condition mean here

Reply:
This code block implements essentially the same logic as `ResumeProducing`, which @westonpace could likely explain better than myself.